src/lib/optimizer/strategy/chunk_pruning_rule.cpp

```diff
@@ -21,150 +22,156 @@
-      std::static_pointer_cast<StoredTableNode>(stored_table_node->deep_copy());
-  stored_table_node_without_column_pruning->set_pruned_column_ids({});
-  const auto predicate_without_column_pruning = expression_copy_and_adapt_to_different_lqp(
-      predicate, {{stored_table_node, stored_table_node_without_column_pruning}});
-  const auto operator_predicates = OperatorScanPredicate::from_expression(*predicate_without_column_pruning,
-                                                                          *stored_table_node_without_column_pruning);
-  // End of hacky
-
-  if (!operator_predicates) return {};
-
-  std::set<ChunkID> result;
-
-  for (const auto& operator_predicate : *operator_predicates) {
-    // Cannot prune column-to-column predicates, at the moment. Column-to-placeholder predicates are never prunable.
-    if (!is_variant(operator_predicate.value)) {
-      continue;
+    // (3.1) Determine set of pruned chunks per predicate chain
+    auto table = Hyrise::get().storage_manager.get_table(stored_table_node->table_name);
+    std::vector<std::set<ChunkID>> pruned_chunk_id_sets;
+    for (auto& predicate_chain : predicate_chains) {
+      auto exclusions = _compute_exclude_list(*table, predicate_chain, stored_table_node);
+      pruned_chunk_id_sets.emplace_back(exclusions);
     }
 
-    const auto column_data_type =
-        stored_table_node_without_column_pruning->output_expressions()[operator_predicate.column_id]->data_type();
+    // (3.2) Calculate the intersect of pruned chunks across predicate chains
+    std::set<ChunkID> pruned_chunk_ids = _intersect_chunk_ids(pruned_chunk_id_sets);
+    if (pruned_chunk_ids.empty()) continue;
+
+    // (3.2) Set pruned chunk ids
+    DebugAssert(stored_table_node->pruned_chunk_ids().empty(),
+                "Did not expect a StoredTableNode with an already existing set of pruned chunk ids.");
+    // Wanted side effect of using sets: pruned_chunk_ids vector is already sorted
+    stored_table_node->set_pruned_chunk_ids(std::vector<ChunkID>(pruned_chunk_ids.begin(), pruned_chunk_ids.end()));
+  }
+}
 
-    // If `value` cannot be converted losslessly to the column data type, we rather skip pruning than running into
-    // errors with lossful casting and pruning Chunks that we shouldn't have pruned.
-    auto value = lossless_variant_cast(boost::get<AllTypeVariant>(operator_predicate.value), column_data_type);
-    if (!value) {
+std::set<ChunkID> ChunkPruningRule::_compute_exclude_list(
+    const Table& table, const PredicateChain& predicate_chain,
+    const std::shared_ptr<StoredTableNode>& stored_table_node) const {
+  std::set<ChunkID> global_excluded_chunk_ids;
+  for (const auto& predicate_node : predicate_chain) {
+    /**
+     * Determine the set of chunks that can be excluded for the given PredicateNode's predicate.
+     */
+    std::set<ChunkID> local_excluded_chunk_ids;
+
+    auto excluded_chunk_ids_iter = _excluded_chunk_ids_by_predicate_node.find(predicate_node);
```

Comment:
I am not a fan of passing state around in both the arguments and the object. If this is solely a cache, it is somewhat fine, but "cache" should be part of its name